pkg/apis/events/validation/validation.go

```diff
@@ -0,0 +1,84 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	events "k8s.io/api/events/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	reportingInstanceLengthLimit = 128
+	actionLengthLimit            = 128
+	reasonLengthLimit            = 128
+	noteLengthLimit              = 1024
+)
+
+// ValidateEvent makes sure that the event makes sense.
+func ValidateEvent(event *events.Event) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// Because go
+	zeroTime := time.Time{}
+
+	if event.EventTime.Time == zeroTime {
+		allErrs = append(allErrs, field.Required(field.NewPath("eventTime"), ""))
+	}
+	if len(event.Regarding.Namespace) == 0 && event.Namespace != metav1.NamespaceSystem {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("regarding", "namespace"), event.Regarding.Namespace, "does not match event.namespace"))
+	}
+	if len(event.ReportingController) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reportingController"), ""))
+	}
+	for _, msg := range validation.IsQualifiedName(event.ReportingController) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reportingController"), event.ReportingController, msg))
+	}
+	if len(event.ReportingInstance) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reportingInstance"), ""))
+	}
+	if len(event.ReportingInstance) > reportingInstanceLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reportingInstance"), "", fmt.Sprintf("can have at most %v characters", reportingInstanceLengthLimit)))
+	}
+	if len(event.Action) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("action"), ""))
+	}
+	if len(event.Action) > actionLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("action"), "", fmt.Sprintf("can have at most %v characters", actionLengthLimit)))
+	}
+	if len(event.Reason) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("reason"), ""))
+	}
+	if len(event.Reason) > reasonLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("reason"), "", fmt.Sprintf("can have at most %v characters", reasonLengthLimit)))
+	}
+	if len(event.Note) > noteLengthLimit {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("note"), "", fmt.Sprintf("can have at most %v characters", noteLengthLimit)))
+	}
+	if event.Type != v1.EventTypeNormal && event.Type != v1.EventTypeWarning {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("type"), "", fmt.Sprintf("has invalid value: %v", event.Type)))
+	}
+
+	for _, msg := range validation.IsDNS1123Subdomain(event.Namespace) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("namespace"), event.Namespace, msg))
+	}
+	return allErrs
```

Comment:
Can we also add validation that the deprecated fields (there are 4 of them) are actually unset?

Reply:
Done.